### PR TITLE
GRAMEX-67 ⁃ InfluxDB for FormHandler

### PR DESCRIPTION
**Pull request by jaidevd**
### InfluxDB setup

Run InfluxDB with Docker:
```bash
$ docker pull influxdb:2.0.8
$ docker run docker run --name influxdb -p 8086:8086 influxdb:2.0.8
```
Then go to [http://localhost:8086](http://localhost:8086) and do the initial setup. Create the following:

1. A username / password pair
2. An org
3. A bucket

Then, got to the "Data" menu on the left pane on the InfluxDB UI:

Screenshot from 2021-09-01 10-09-06

Then, in the "Buckets" tab, create a bucket:

Screenshot from 2021-09-01 10-10-42

Finally, in the "Tokens" tab, copy the token for your username:

image

### Gramex Setup

```yaml
url:
  home:
    pattern: /$YAMLURL/
    handler: FormHandler
    kwargs:
      url: influxdb:http://localhost:8086/
      username: jaidevd
      token: $token  # Token copied from earlier
      org: Gramener
      bucket: stocks
      xsrf_cookies: false
      id: ''
```

### Adding Data

Sample:

```json
{
  "symbol": [
    "AAPL",
    "GOOG",
    "MSFT"
  ],
  "open": [
    213.429998,
    626.951088,
    30.620001
  ],
  "close": [
    214.009998,
    626.751061,
    30.950001
  ],
  "low": [
    212.380001,
    624.241073,
    30.59
  ],
  "high": [
    214.499996,
    629.511067,
    31.1
  ],
  "volume": [
    123432400,
    3927000,
    38409100
  ],
  "_time": [
    "2014-09-05 15:48:19.162191",
    "2014-09-05 15:48:19.162191",
    "2014-09-05 15:48:19.162191"
  ]
}
```
Download a bigger file [here](https://gramener0-my.sharepoint.com/:u:/g/personal/jaidev_deshpande_gramener_com/ES1akeukSIRLu23JcuB4u1AB0hm-WErWitApgGpWTNvbXw?e=BPbyY8).

Post data in bulk with with
```bash
POST -H 'Content-Type: application/json' --data '@stocks.json' http://localhost:9988/\?measurement\=stocks\&tags\=symbol
```

Post individual records as:

```bash
POST http://localhost:9988/?measurement=stocks&tags=symbol&symbol=AAPL&high=1&low=2&volume=3&open=4&close=5
```
`_time` is optional. If not specified, will be considered as the server time.

### Querying data
```bash

# Get values of the Apple stock in the last one week:
GET http://localhost:9988/?bucket=stocks&symbol=AAPL&_offset=-7d

# Get only volume values
GET http://localhost:9988/?bucket=stocks&symbol=AAPL&_offset=-7d&_c=volume

# Get everything except the volume
GET http://localhost:9988/?bucket=stocks&symbol=AAPL&_offset=-7d&_c=-volume

# Get Apple and Google stocks on the days when they closed at > 100 points.
GET http://localhost:9988/?bucket=stocks&symbol=AAPL&symbol=GOOG&close>=100
```

┆Issue is synchronized with this [Jira Bug](https://gramenertech.atlassian.net/browse/GRAMEX-67)

### CODE REVIEW

File: gramex/data.py [hunk 1]
```diff
@@ -1498,6 +1499,69 @@ def _insert_mongodb(url, rows, meta=None, database=None, collection=None, **kwar
     return len(result.inserted_ids)
 
 
+def _filter_influxdb(url, controls, args, org=None, bucket=None, query=None, **kwargs):
+    with _influxdb_client(url, org=org, **kwargs) as db:
+        q = db.query_api()
+        _range = args.pop('range')[0]
```
Comment (sanand0) [inline gramex/data.py:1505]: Use _offset and _limit instead of range. Handle time conversion as required.

File: gramex/data.py [hunk 2]
```diff
@@ -1498,6 +1499,69 @@ def _insert_mongodb(url, rows, meta=None, database=None, collection=None, **kwar
...
+    rows = _get_ts_points(rows, measurement, tags)
+    from influxdb_client.client.write_api import ASYNCHRONOUS
+    with _influxdb_client(url, **kwargs) as db:
+        with db.write_api(write_options=ASYNCHRONOUS) as client:
+            client.write(
+                bucket=bucket, org=db.org, record=rows,
+                data_frame_measurement_name=measurement,
+                data_frame_tag_columns=tags
+            ).get()
+    return len(rows)
```
Comment (sanand0) [inline gramex/data.py:1562]: Set `meta['inserted'] = [array of {'id': <id>}]`

File: gramex/data.py (hunk 1 above)
Comment (sanand0) [reply gramex/data.py:1505]: We want support for:

```text
?Continent=Europe ► Continent = Europe
?Continent=Europe&Continent=Asia ► Continent = Europe OR Asia. Multiple values are allowed
?Continent!=Europe ► Continent is NOT Europe
?Continent!=Europe&Continent!=Asia ► Continent is NEITHER Europe NOR Asia
OPTIONAL: ?Shapes ► Shapes is not NULL
OPTIONAL: ?Shapes! ► Shapes is NULL
?c1>=10 ► c1 > 10 (not >= 10)
?c1>~=10 ► c1 >= 10. The ~ acts like an =
?c1<=10 ► c1 < 10 (not <= 10)
?c1<~=10 ► c1 <= 10. The ~ acts like an =
?c1>=10&c1<=20 ► c1 > 10 AND c1 < 20
OPTIONAL: ?Name~=United ► Name matches &_format=html
OPTIONAL: ?Name!~=United ► Name does NOT match United
OPTIONAL: ?Name~=United&Continent=Asia ► Name matches United AND Continent is Asia
```

To control the output, you can use these control arguments:

```text
Limit rows: ?_limit=10 ► show only 10 rows
Offset rows: ?_offset=10&_limit=10 ► show only 10 rows starting, skipping the first 10 rows
OPTIONAL: Sort by columns: ?_sort=Continent&_sort=Name ► sort first by Continent (ascending) then Name (ascending)
OPTIONAL: Sort order: ?_sort=-Continent&_sort=-ID ► sort first by Continent (descending) then ID (descending)
OPTIONAL: Specific columns: ?_c=Continent&_c=Name ► show only the Continent and Names columns
OPTIONAL: Exclude columns: ?_c=-Continent&_c=-Name ► show all columns except the Continent and Names columns
```

Use `_pop_controls()`, `_filter_col()`, maybe `_filter_sort_columns()` from data.py in your code.

**Comment by jaidevd** [comment]
@sanand0 This is ready. Whatever is mentioned in the PR is implemented.

**Pull request closed without merging**